README.md
# MetaheuristicAlgorithmsPython

Various metaheuristic algorithms implemented in Python.

This is equivalent to MetaheuristicAlgorithms written in Ruby (https://github.com/tadatoshi/metaheuristic_algorithms). The reason why I wrote it in Python is that I would like to potentially utilize Python's Scientific Computing libraries. 

As a programming lanugage, I prefer Ruby, because it's fully Object-Oriented programming language (also Dynamic language) and because it has a community with the culture of writing unit tests. Both of these characteristics lead to cleaner, well structured, easy-to-maintain codes. Also it's easier to understand other people's codes written in such a way. 

But scientists use Python for their activities such as Scientific Computing, Optimization, Data Science, Data Mining, Machine Learning etc. In other words, Python has a community of scientists.  

## Installation

Use ``pip3``:

```
pip3 install metaheuristic_algorithms_python
```

## Supported Platforms

* Python 3.5. 

It's not tested on Python 2.6 or 2.7 yet. 

## Available Algorithms

* Harmony Search

* Simplified Particle Swarm Optimization

* Simulated Annealing

* Firefly Algorithm

Algorithms under construction:

* Genetic Algorithm

## Provided Objective Functions

Function Wrappers (See "Usage" section below) for the following Objective Functions are available in metaheuristic_algorithms.function_wrappers package

(You can write your own Fuction Wrapper for your Objective Function. See "Usage" section below.)

* Nonsmooth Multipeak Function

* Easom Function

* Rosenbrook Function

* Michaelwicz Function

## Usage

### Programming

Step 1. Create a Function Wrapper for your objective function by extending MetaheuristicAlgorithms::FunctionWrappers::AbstractWrapper

   Example: Rosenbrook's Function: f(x,y) = (1 - x)^2 + 100(y - x^2)^2

```python
    from metaheuristic_algorithms.function_wrappers.abstract_wrapper import AbstractWrapper

    class RosenbrookFunctionWrapper(AbstractWrapper):

        def maximum_decision_variable_values(self):
            return [5, 5]

        def minimum_decision_variable_values(self):
            return [-5, -5]

        def objective_function_value(self, decision_variable_values):
            return (1 - decision_variable_values[0])**2 + 100 * (decision_variable_values[1] - decision_variable_values[0]**2)**2

        def initial_decision_variable_value_estimates(self):
            return [2, 2]
```

Step 2. Instantiate the created Function Wrapper and pass it as the first argument of the Algorithm instantiation. 
        Also specify the number of variables and objective ("maximization" or "minimization")
        Then call the search method passing the algorithm specific parameters. 

   Example: Harmony Search for the glocal minimum value for Rosenbrook's Function

```python
    from metaheuristic_algorithms.harmony_search import HarmonySearch
    from metaheuristic_algorithms.function_wrappers.rosenbrook_function_wrapper import RosenbrookFunctionWrapper

    rosenbrook_function_wrapper = RosenbrookFunctionWrapper()

    number_of_variables = 2
    objective = "minimization"

    harmony_search = HarmonySearch(rosenbrook_function_wrapper, number_of_variables, objective)

    maximum_attempt = 25000
    pitch_adjusting_range = 100
    harmony_search_size = 20
    harmony_memory_acceping_rate = 0.95
    pitch_adjusting_rate = 0.7

    result = harmony_search.search(maximum_attempt = maximum_attempt, 
                                   pitch_adjusting_range = pitch_adjusting_range, 
                                   harmony_search_size = harmony_search_size, 
                                   harmony_memory_acceping_rate = harmony_memory_acceping_rate, 
                                   pitch_adjusting_rate = pitch_adjusting_rate)

    print(result["best_decision_variable_values"][0]) # x value: Example: 1.0112
    print(result["best_decision_variable_values"][1]) # y value: Example: 0.9988
    print(result["best_objective_function_value"])    # f(x,y) value: Example: 0.0563    
```

### Command line execution

Examples: 

* Firefly Algorithm:
```
    $ metaheuristic_algorithms_python --objective_function 'nonsmooth_multipeak_function' --number_of_variables 2 --objective 'maximization' firefly_algorithm --number_of_fireflies 10 --maximun_generation 10 --randomization_parameter_alpha 0.2 --absorption_coefficient_gamma 1.0
```    

* Harmony Search:
```
    $ metaheuristic_algorithms_python --objective_function 'rosenbrook_function' --number_of_variables 2 --objective 'minimization' harmony_search --maximum_attempt 25000 --pitch_adjusting_range 100 --harmony_search_size 20 --harmony_memory_acceping_rate 0.95 --pitch_adjusting_rate 0.7
```    

* Simplified Particle Swarm Optimization:
```
    $ metaheuristic_algorithms_python --objective_function 'michaelwicz_function' --number_of_variables 2 --objective 'minimization' simplified_particle_swarm_optimization --number_of_particiles 20 --number_of_iterations 15 --social_coefficient 0.5 --random_variable_coefficient 0.2
```    

* Simulated Annealing:
```
    $ metaheuristic_algorithms_python --objective_function 'rosenbrook_function' --number_of_variables 2 --objective 'minimization' simulated_annealing --temperature 1.0 --minimal_temperature 1e-10 --maximum_number_of_rejections 2500 --maximum_number_of_runs 500 --maximum_number_of_acceptances 15 --bolzmann_constant 1 --cooling_factor 0.95 --energy_norm 1e-8 --standard_diviation_for_estimation 1 --ratio_of_energy_delta_over_evaluation_delta 1
```

## Development



## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/tadatoshi/metaheuristic_algorithms_python. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](contributor-covenant.org) code of conduct.


## License

The project is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).

## References

X.-S. Yang, Engineering Optimization : An Introduction with Metaheuristic Applications. Hoboken, New Jersey, United States: John Wiley & Sons, 2010.



CHANGELOG.md
## metaheuristic_algorithms_python 0.1.5 (November 22, 2015) ##

* Added command line script. 

## metaheuristic_algorithms_python 0.1.4 (November 21, 2015) ##

* Fixed packages in setup.py so that metaheuristic_algorithms_python.function_wrappers subpackage is included. 

## metaheuristic_algorithms_python 0.1.3 (November 12, 2015) ##

* Added Firefly Algorithm

## metaheuristic_algorithms_python 0.1.2 (November 9, 2015) ##

* Fixed a typo in method name minimum_decision_variable_values in FunctionWrapper's.

* Added Simulated Annealing

## metaheuristic_algorithms_python 0.1.1 (November 3, 2015) ##

* Added Simplified Particle Swarm Optimization

## metaheuristic_algorithms_python 0.1.0 (October 26, 2015) ##

* Added Harmony Search


metaheuristic_algorithms/command_line.py
import argparse

def main():

    '''
    Based on https://docs.python.org/3.5/library/argparse.html#sub-commands

    Examples
    -----------

    Firefly Algorithm:
    $ metaheuristic_algorithms_python --objective_function 'nonsmooth_multipeak_function' --number_of_variables 2 --objective 'maximization' firefly_algorithm --number_of_fireflies 10 --maximun_generation 10 --randomization_parameter_alpha 0.2 --absorption_coefficient_gamma 1.0

    Harmony Search:
    $ metaheuristic_algorithms_python --objective_function 'rosenbrook_function' --number_of_variables 2 --objective 'minimization' harmony_search --maximum_attempt 25000 --pitch_adjusting_range 100 --harmony_search_size 20 --harmony_memory_acceping_rate 0.95 --pitch_adjusting_rate 0.7

    Simplified Particle Swarm Optimization:
    $ metaheuristic_algorithms_python --objective_function 'michaelwicz_function' --number_of_variables 2 --objective 'minimization' simplified_particle_swarm_optimization --number_of_particiles 20 --number_of_iterations 15 --social_coefficient 0.5 --random_variable_coefficient 0.2

    Simulated Annealing:
    $ metaheuristic_algorithms_python --objective_function 'rosenbrook_function' --number_of_variables 2 --objective 'minimization' simulated_annealing --temperature 1.0 --minimal_temperature 1e-10 --maximum_number_of_rejections 2500 --maximum_number_of_runs 500 --maximum_number_of_acceptances 15 --bolzmann_constant 1 --cooling_factor 0.95 --energy_norm 1e-8 --standard_diviation_for_estimation 1 --ratio_of_energy_delta_over_evaluation_delta 1
    '''

    parser = argparse.ArgumentParser(prog='Executes a specified metaheuristic algorithm')

    parser.add_argument('--objective_function', nargs='?', required=True, choices=['easom_function', 'michaelwicz_function', 'nonsmooth_multipeak_function', 'rosenbrook_function'])
    parser.add_argument('--number_of_variables', nargs='?', type=int, required=True)
    parser.add_argument('--objective', nargs='?', required=True, choices=['maximization', 'minimization'])    

    # "dest": the name of the variable that holds the name of subparser.
    subparsers = parser.add_subparsers(title='algorithm_name', dest='algorithm_name')

    parser_firefly_algorithm = subparsers.add_parser('firefly_algorithm')
    parser_firefly_algorithm.add_argument('--number_of_fireflies', nargs='?', type=int, required=True)
    parser_firefly_algorithm.add_argument('--maximun_generation', nargs='?', type=int, required=True)
    parser_firefly_algorithm.add_argument('--randomization_parameter_alpha', nargs='?', type=float, required=True)
    parser_firefly_algorithm.add_argument('--absorption_coefficient_gamma', nargs='?', type=float, required=True)
    # Note: Calls execute_firefly_algorithm function with the arguments:
    parser_firefly_algorithm.set_defaults(func=execute_firefly_algorithm)

    parser_harmony_search = subparsers.add_parser('harmony_search')
    parser_harmony_search.add_argument('--maximum_attempt', nargs='?', type=int, required=True)
    parser_harmony_search.add_argument('--pitch_adjusting_range', nargs='?', type=int, required=True)
    parser_harmony_search.add_argument('--harmony_search_size', nargs='?', type=int, required=True)
    parser_harmony_search.add_argument('--harmony_memory_acceping_rate', nargs='?', type=float, required=True)
    parser_harmony_search.add_argument('--pitch_adjusting_rate', nargs='?', type=float, required=True)
    parser_harmony_search.set_defaults(func=execute_harmony_search)

    parser_simplified_particle_swarm_optimization = subparsers.add_parser('simplified_particle_swarm_optimization')
    parser_simplified_particle_swarm_optimization.add_argument('--number_of_particiles', nargs='?', type=int, required=True)
    parser_simplified_particle_swarm_optimization.add_argument('--number_of_iterations', nargs='?', type=int, required=True)
    parser_simplified_particle_swarm_optimization.add_argument('--social_coefficient', nargs='?', type=float, required=True)
    parser_simplified_particle_swarm_optimization.add_argument('--random_variable_coefficient', nargs='?', type=float, required=True)
    parser_simplified_particle_swarm_optimization.set_defaults(func=execute_simplified_particle_swarm_optimization)

    parser_simulated_annealing = subparsers.add_parser('simulated_annealing')
    parser_simulated_annealing.add_argument('--temperature', nargs='?', type=float, required=True)
    parser_simulated_annealing.add_argument('--minimal_temperature', nargs='?', type=float, required=True)
    parser_simulated_annealing.add_argument('--maximum_number_of_rejections', nargs='?', type=int, required=True)
    parser_simulated_annealing.add_argument('--maximum_number_of_runs', nargs='?', type=int, required=True)
    parser_simulated_annealing.add_argument('--maximum_number_of_acceptances', nargs='?', type=int, required=True)
    parser_simulated_annealing.add_argument('--bolzmann_constant', nargs='?', type=float, required=True)
    parser_simulated_annealing.add_argument('--cooling_factor', nargs='?', type=float, required=True)
    parser_simulated_annealing.add_argument('--energy_norm', nargs='?', type=float, required=True)
    parser_simulated_annealing.add_argument('--standard_diviation_for_estimation', nargs='?', type=float, required=True)
    parser_simulated_annealing.add_argument('--ratio_of_energy_delta_over_evaluation_delta', nargs='?', type=float, required=True)
    parser_simulated_annealing.set_defaults(func=execute_simulated_annealing)

    args = parser.parse_args()
    # Calls the function specified in "set_defaults" method:
    args.func(args)

def execute_firefly_algorithm(args):
    objective_function = instantiate_objective_function(args.objective_function)

    from metaheuristic_algorithms.firefly_algorithm import FireflyAlgorithm
    firefly_algorithm = FireflyAlgorithm(objective_function, args.number_of_variables, args.objective)

    result = firefly_algorithm.search(number_of_fireflies = args.number_of_fireflies, maximun_generation = args.maximun_generation, 
                                      randomization_parameter_alpha = args.randomization_parameter_alpha, absorption_coefficient_gamma = args.absorption_coefficient_gamma)

    print('best_decision_variable_values=', result["best_decision_variable_values"])
    print('best_objective_function_value=', result["best_objective_function_value"])

def execute_harmony_search(args):
    objective_function = instantiate_objective_function(args.objective_function)

    from metaheuristic_algorithms.harmony_search import HarmonySearch
    harmony_search = HarmonySearch(objective_function, args.number_of_variables, args.objective)

    result = harmony_search.search(maximum_attempt = args.maximum_attempt, pitch_adjusting_range = args.pitch_adjusting_range, 
                                   harmony_search_size = args.harmony_search_size, harmony_memory_acceping_rate = args.harmony_memory_acceping_rate, 
                                   pitch_adjusting_rate = args.pitch_adjusting_rate)

    print('best_decision_variable_values=', result["best_decision_variable_values"])
    print('best_objective_function_value=', result["best_objective_function_value"])

def execute_simplified_particle_swarm_optimization(args):
    objective_function = instantiate_objective_function(args.objective_function)

    from metaheuristic_algorithms.simplified_particle_swarm_optimization import SimplifiedParticleSwarmOptimization
    simplified_particle_swarm_optimization = SimplifiedParticleSwarmOptimization(objective_function, args.number_of_variables, args.objective)

    result = simplified_particle_swarm_optimization.search(number_of_particiles = args.number_of_particiles, number_of_iterations = args.number_of_iterations, 
                                                           social_coefficient = args.social_coefficient, random_variable_coefficient = args.random_variable_coefficient)

    print('best_decision_variable_values=', result["best_decision_variable_values"])
    print('best_objective_function_value=', result["best_objective_function_value"])

def execute_simulated_annealing(args):
    objective_function = instantiate_objective_function(args.objective_function)

    from metaheuristic_algorithms.simulated_annealing import SimulatedAnnealing
    simulated_annealing = SimulatedAnnealing(objective_function, args.number_of_variables, args.objective)

    result = simulated_annealing.search(args.temperature, args.minimal_temperature, 
                                        args.bolzmann_constant, args.energy_norm, 
                                        maximum_number_of_rejections = args.maximum_number_of_rejections, 
                                        maximum_number_of_runs = args.maximum_number_of_runs, 
                                        maximum_number_of_acceptances = args.maximum_number_of_acceptances, 
                                        cooling_factor = args.cooling_factor, 
                                        standard_diviation_for_estimation = args.standard_diviation_for_estimation, 
                                        ratio_of_energy_delta_over_evaluation_delta = args.ratio_of_energy_delta_over_evaluation_delta)

    print('best_decision_variable_values=', result["best_decision_variable_values"])
    print('best_objective_function_value=', result["best_objective_function_value"])

def instantiate_objective_function(objective_function_name):
    if objective_function_name == 'easom_function':
        from metaheuristic_algorithms.function_wrappers.easom_function_wrapper import EasomFunctionWrapper
        return EasomFunctionWrapper()
    elif objective_function_name == 'michaelwicz_function':
        from metaheuristic_algorithms.function_wrappers.michaelwicz_function_wrapper import MichaelwiczFunctionWrapper
        return MichaelwiczFunctionWrapper()
    elif objective_function_name == 'nonsmooth_multipeak_function':
        from metaheuristic_algorithms.function_wrappers.nonsmooth_multipeak_function_wrapper import NonsmoothMultipeakFunctionWrapper
        return NonsmoothMultipeakFunctionWrapper()
    elif objective_function_name == 'rosenbrook_function':
        from metaheuristic_algorithms.function_wrappers.rosenbrook_function_wrapper import RosenbrookFunctionWrapper
        return RosenbrookFunctionWrapper()
    else:
        return None

if __name__ == "__main__":
    main() 